### PR TITLE
Filter generate types

### PR DESCRIPTION
**Pull request by nitzan-haim**
Added an optional field to the apiDefinitions in the config of adapters that use the adapter-components infrastructure with **swagger**.
It could be used to specify the endpoints supported by salto, and only the types based on those endpoints will be generated from the yaml.

---

Note that this new config field will be hidden once the whole `apiDefinitions` part of the config will be hidden

---

*Release Notes*: None

### CODE REVIEW

**Comment by ori-moisis** [comment]
@alonstern - is this similar / duplicate to the logic you added in #2043 ?

**Comment by alonstern** [comment]
> @alonstern - is this similar / duplicate to the logic you added in #2043 ?

I don't think so. Here the filter is by the URL endpoints, in my PR the filter is by types after all the types were generated

File: packages/adapter-components/src/config/swagger.ts [hunk 1]
```diff
@@ -53,6 +53,7 @@ export type TypeSwaggerDefaultConfig = TypeDefaultsConfig
 
 export type AdapterSwaggerApiConfig = AdapterApiConfig & {
   swagger: SwaggerDefinitionBaseConfig
+  supportedEndpoints?: string[]
```
Comment (netama) [inline packages/adapter-components/src/config/swagger.ts:56]: let's call this `supportedTypes` for consistency (and compare the `toTypeName()` values)? this should use the same terminology as `includeTypes`

File: packages/adapter-components/test/elements/swagger/type_elements.test.ts [hunk 2]
```diff
@@ -279,6 +279,25 @@ describe('swagger_type_elements', () => {
...
+          {
+            swagger: { url: `${BASE_DIR}/petstore_swagger.v2.yaml` },
+            typeDefaults: { transformation: { idFields: ['name'] } },
+            types: {},
+            supportedEndpoints: ['/pet/{petId}'],
+          },
+        )
+        allTypes = res.allTypes
+      })
+      it('should generate the right types', () => {
```
Comment (netama) [inline packages/adapter-components/test/elements/swagger/type_elements.test.ts:296]: nit: maybe make the description a little more specific to this case (something like `should generate only the types included in supportedTypes and their subtypes`)?

File: packages/adapter-components/test/elements/swagger/type_elements.test.ts [hunk 3]
```diff
@@ -279,6 +279,25 @@ describe('swagger_type_elements', () => {
...
+        expect(Object.keys(allTypes).sort()).toEqual(['Pet', 'Category', 'Tag'].sort())
+      })
+    })
+
```
Comment (netama) [inline packages/adapter-components/test/elements/swagger/type_elements.test.ts:300]: may be worth also testing this somewhere in combination with the instance fetch logic? something like valid+invalid scenarios under `instance_elements.test.ts` (at least for the [swagger](https://github.com/salto-io/salto/blob/master/packages/adapter-components/test/elements/swagger/instance_elements.test.ts) variant, maybe also the ducktype one)?

**Review by netama — COMMENTED** [review]
Looks good :) 
IIRC we talked originally about adding `supportedTypes` which could also be used as a reference for the `includeTypes`, was there a reason for the switch to `endpoint`? 
Also - in cases where we define `supportedTypes`, wdyt about adding a config validation that ensures `includeTypes` only includes types from `supportedTypes` (similar to the `ducktype` validation [here](https://github.com/salto-io/salto/blob/master/packages/adapter-components/src/config/ducktype.ts#L74))?

File: packages/adapter-components/src/config/swagger.ts (hunk 1 above)
Comment (netama) [reply packages/adapter-components/src/config/swagger.ts:56]: also - if we're keeping this in the shared config let's also add support for it in the `ducktype` variant (it won't be super useful, but it should probably be either a config validation with the `types` mapping that I linked in the other comment, or a filter in the type-generation logic like what you added in the swagger)? if you prefer to move this inside the `swagger` config instead that also works

**Comment by netama** [comment]
@nitzan-haim @alonstern @ori-moisis re the `subtypes` variant in https://github.com/salto-io/salto/pull/2043 - I agree with Ori that we should pick one of the implementations (doesn't really matter if filtering is done during type generation or after it), but IIUC the intended usage is a little different and we should agree on which one we want (and keep it consistent between swagger and soap unless there's a reason not to?) - Nitzan's feature addition allows explicitly specifying all the supported types, separately from those currently included in the fetch (the former is supposed to be defined by us when implementing the adapter, and the latter is part of the user configuration). From what I understand Alon's use case is still a little different - maybe we can go over it together to understand which makes more sense to add (and if the use cases are different enough maybe this should be a swagger-only config after all)?

**Comment by nitzan-haim** [comment]
closing due to a decision to change the solution, and rebase of lazy elements

**Pull request closed without merging**